### PR TITLE
fix: fix event handlers implicit any type inference

### PR DESCRIPTION
**Pull request by zchenwei**
#### Description of changes

- Moving `IsAny<Element>` conditional type check a level up before merging HTML element's attributes to ensure that Typescript will do an additional inference phase for the generics `Element`.
- Fix some primitive prop types

This will fix the type issue that occurs on event handlers in cases where people try to pass an inlined function but the type of `event` parameter is implicitly inferred as `any`. For example:

Before:
`e` was inferred as `any`
Screenshot 2023-07-05 at 12 55 59 PM

After:
`e` was inferred as a specific type
Screenshot 2023-07-05 at 12 27 17 PM

#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added
- [x] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ioanabrooks — DISMISSED** [review]
Pulled branch and tested. LGTM!

**Comment by zchenwei** [comment]
> Pulled branch and tested. LGTM!

Thanks! I am adding a changeset rn.

File: .changeset/smart-mails-accept.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"@aws-amplify/ui-docs": patch
+"@aws-amplify/ui-react": patch
```
Comment (ioanabrooks) [inline .changeset/smart-mails-accept.md]:
```suggestion
"@aws-amplify/ui-react": patch
```

Comment (ioanabrooks) [reply .changeset/smart-mails-accept.md]: No patch for docs ;)

Comment (zchenwei) [reply .changeset/smart-mails-accept.md]: I deleted one sec ago 😅

**Review by ioanabrooks — APPROVED** [review]

**Review by reesscot — APPROVED** [review]

**Review by ioanabrooks — APPROVED** [review]

**Review by wlee221 — APPROVED** [review]

**Pull request merged**